tests/types/test_mxc_uri.py

```diff
@@ -0,0 +1,94 @@
+        mxc_0 = MXCUri(server_name="example.com", media_id="84n8493hnfsjkbcu")
+        self.assertEqual(str(mxc_0), "mxc://example.com/84n8493hnfsjkbcu")
+
+        mxc_1 = MXCUri(
+            server_name="192.168.1.17:8008", media_id="bajkad89h31ausdhoqqasd"
+        )
+        self.assertEqual(str(mxc_1), "mxc://192.168.1.17:8008/bajkad89h31ausdhoqqasd")
+
+        mxc_2 = MXCUri(server_name="123.123.123.123", media_id="000000000000")
+        self.assertEqual(str(mxc_2), "mxc://123.123.123.123/000000000000")
+
+    def test_valid_mxc_uris_from_str(self) -> None:
+        """Tests that a series of valid mxc uris strs are parsed correctly."""
+        # Converting a str to its MXCUri representation
+        mxcuri_0 = MXCUri.from_str("mxc://example.com/g12789g890ajksjk")
+        self.assertEqual(mxcuri_0.server_name, "example.com")
+        self.assertEqual(mxcuri_0.media_id, "g12789g890ajksjk")
+
+        mxcuri_1 = MXCUri.from_str("mxc://localhost:8448/abcdefghijklmnopqrstuvwxyz")
+        self.assertEqual(mxcuri_1.server_name, "localhost:8448")
+        self.assertEqual(mxcuri_1.media_id, "abcdefghijklmnopqrstuvwxyz")
+
+        mxcuri_2 = MXCUri.from_str("mxc://[::1]/abcdefghijklmnopqrstuvwxyz")
+        self.assertEqual(mxcuri_2.server_name, "[::1]")
+        self.assertEqual(mxcuri_2.media_id, "abcdefghijklmnopqrstuvwxyz")
+
+        mxcuri_3 = MXCUri.from_str("mxc://123.123.123.123:32112/12893y81283781023")
+        self.assertEqual(mxcuri_3.server_name, "123.123.123.123:32112")
+        self.assertEqual(mxcuri_3.media_id, "12893y81283781023")
+
+        mxcuri_4 = MXCUri.from_str("mxc://domain/abcdefg")
+        self.assertEqual(mxcuri_4.server_name, "domain")
+        self.assertEqual(mxcuri_4.media_id, "abcdefg")
+
+    def test_invalid_mxc_uris_from_str(self) -> None:
+        """Tests that a series of invalid mxc uris are appropriately rejected."""
+        # Converting invalid MXC URI strs to MXCUri representations
+        with self.assertRaises(ValueError):
+            MXCUri.from_str("http://example.com/abcdef")
+
+        with self.assertRaises(ValueError):
+            MXCUri.from_str("mxc:///example.com/abcdef")
+
+        with self.assertRaises(ValueError):
+            MXCUri.from_str("mxc://example.com//abcdef")
+
+        with self.assertRaises(ValueError):
+            MXCUri.from_str("mxc://example.com/abcdef/")
+
+        with self.assertRaises(ValueError):
+            MXCUri.from_str("mxc://example.com/abc/abcdef")
+
+        with self.assertRaises(ValueError):
+            MXCUri.from_str("mxc://example.com/abc/abcdef")
+
+        with self.assertRaises(ValueError):
+            MXCUri.from_str("mxc:///abcdef")
+
+        with self.assertRaises(ValueError):
+            MXCUri.from_str("mxc://example.com")
+
+        with self.assertRaises(ValueError):
+            MXCUri.from_str("mxc://example.com/")
+
+        with self.assertRaises(ValueError):
+            MXCUri.from_str("mxc:///")
```

Comment:
Probably worth trying a scheme-less one since we found that has slightly weird behavior:

```suggestion
        with self.assertRaises(ValueError):
            MXCUri.from_str("mxc:///")

        with self.assertRaises(ValueError):
            MXCUri.from_str("example.com/abc")
```